README.md
# conda-mirror

[![Project Status: Inactive – The project has reached a stable, usable state but is no longer being actively developed; support/maintenance will be provided as time allows.](https://www.repostatus.org/badges/latest/inactive.svg)](https://www.repostatus.org/#inactive)
[![Build Status](https://travis-ci.org/Valassis-Digital-Media/conda-mirror.svg?branch=master)](https://travis-ci.org/Valassis-Digital-Media/conda-mirror)
[![PyPI version](https://badge.fury.io/py/conda-mirror.svg)](https://badge.fury.io/py/conda-mirror)
[![codecov](https://codecov.io/gh/Valassis-Digital-Media/conda-mirror/branch/master/graph/badge.svg)](https://codecov.io/gh/Valassis-Digital-Media/conda-mirror)

Mirrors an upstream conda channel to a local directory.

## Install

`conda-mirror` is available on PyPI and conda-forge.

Install with:

`pip install conda-mirror`

or:

`conda install conda-mirror -c conda-forge`

## Compatibility

`conda-mirror` is intentionally a py3 only package

## CLI

CLI interface for `conda-mirror.py`

```
usage: conda-mirror [-h] [--upstream-channel UPSTREAM_CHANNEL]
                    [--target-directory TARGET_DIRECTORY]
                    [--temp-directory TEMP_DIRECTORY] [--platform PLATFORM]
                    [-v] [--config CONFIG] [--pdb] [--num-threads NUM_THREADS]
                    [--version] [--dry-run] [--no-validate-target]
                    [--minimum-free-space MINIMUM_FREE_SPACE]

CLI interface for conda-mirror.py

optional arguments:
  -h, --help            show this help message and exit
  --upstream-channel UPSTREAM_CHANNEL
                        The target channel to mirror. Can be a channel on
                        anaconda.org like "conda-forge" or a full qualified
                        channel like "https://repo.continuum.io/pkgs/free/"
  --target-directory TARGET_DIRECTORY
                        The place where packages should be mirrored to
  --temp-directory TEMP_DIRECTORY
                        Temporary download location for the packages. Defaults
                        to a randomly selected temporary directory. Note that
                        you might need to specify a different location if your
                        default temp directory has less available space than
                        your mirroring target
  --platform PLATFORM   The OS platform(s) to mirror. one of: {'linux-64',
                        'linux-32','osx-64', 'win-32', 'win-64'}
  -v, --verbose         logging defaults to error/exception only. Takes up to
                        three '-v' flags. '-v': warning. '-vv': info. '-vvv':
                        debug.
  --config CONFIG       Path to the yaml config file
  --pdb                 Enable PDB debugging on exception
  --num-threads NUM_THREADS
                        Num of threads for validation. 1: Serial mode. 0: All
                        available.
  --version             Print version and quit
  --dry-run             Show what will be downloaded and what will be removed.
                        Will not validate existing packages
  --no-validate-target  Skip validation of files already present in target-
                        directory
  --minimum-free-space MINIMUM_FREE_SPACE
                        Threshold for free diskspace. Given in megabytes.
```

## Example Usage

WARNING: Invoking this command will pull ~10GB and take at least an hour

`conda-mirror --upstream-channel conda-forge --target-directory local_mirror --platform linux-64`

## More Details

### blacklist/whitelist configuration

example-conf.yaml:

```yaml
blacklist:
    - license: "*agpl*"
    - license: None
    - license: ""

whitelist:
    - name: system
```

`blacklist` removes package(s) that match the condition(s) listed from the
upstream repodata.

`whitelist` re-includes any package(s) from blacklist that match the
whitelist conditions.

blacklist and whitelist both take lists of dictionaries. The keys in the
dictionary need to be values in the `repodata.json` metadata. The values are
(unix) globs to match on. Go here for the full repodata of the upstream
"defaults" channel:
http://conda.anaconda.org/anaconda/linux-64/repodata.json

Here are the contents of one of the entries in repodata['packages']

```python
{'botocore-1.4.10-py34_0.tar.bz2': {'arch': 'x86_64',
  'binstar': {'channel': 'main',
   'owner_id': '55fc8527d3234d09d4951c71',
   'package_id': '56b88ea1be1cc95a362b218e'},
  'build': 'py34_0',
  'build_number': 0,
  'date': '2016-04-11',
  'depends': ['docutils >=0.10',
   'jmespath >=0.7.1,<1.0.0',
   'python 3.4*',
   'python-dateutil >=2.1,<3.0.0'],
  'license': 'Apache',
  'md5': 'b35a5c1240ba672e0d9d1296141e383c',
  'name': 'botocore',
  'platform': 'linux',
  'requires': [],
  'size': 1831799,
  'version': '1.4.10'}}
```

See implementation details in the `conda_mirror:match` function for more
information.

#### Common usage patterns
##### Mirror **only** one specific package
If you wanted to match exactly the botocore package listed above with your
config, then you could use the following configuration to first blacklist
**all** packages and then include just the botocore packages:

```yaml
blacklist:
    - name: "*"
whitelist:
    - name: botocore
      version: 1.4.10
      build: py34_0
```
##### Mirror everything but agpl licenses
```yaml
blacklist:
    - license: "*agpl*"
```

##### Mirror only python 3 packages
```yaml
blacklist:
    - name: "*"
whitelist:
    - build: "*py3*"
```

## Testing

### Install test requirements

Note: Will install packages from pip

```
$ pip install -r test-requirements.txt
Requirement already satisfied: pytest in /home/edill/miniconda/lib/python3.5/site-packages (from -r test-requirements.txt (line 1))
Requirement already satisfied: coverage in /home/edill/miniconda/lib/python3.5/site-packages (from -r test-requirements.txt (line 2))
Requirement already satisfied: pytest-ordering in /home/edill/miniconda/lib/python3.5/site-packages (from -r test-requirements.txt (line 3))
Requirement already satisfied: py>=1.4.29 in /home/edill/miniconda/lib/python3.5/site-packages (from pytest->-r test-requirements.txt (line 1))
```

### Run the tests, invoking with the `coverage` tool.

```
$ coverage run run_tests.py
sys.argv=['run_tests.py']
========================================= test session starts ==========================================
platform linux -- Python 3.5.3, pytest-3.0.6, py-1.4.31, pluggy-0.4.0 -- /home/edill/miniconda/bin/python
cachedir: .cache
rootdir: /home/edill/dev/maxpoint/github/conda-mirror, inifile:
plugins: xonsh-0.5.2, ordering-0.4
collected 4 items

test/test_conda_mirror.py::test_match PASSED
test/test_conda_mirror.py::test_cli[https://repo.continuum.io/pkgs/free-linux-64] PASSED
test/test_conda_mirror.py::test_cli[conda-forge-linux-64] PASSED
test/test_conda_mirror.py::test_handling_bad_package PASSED

======================================= 4 passed in 4.41 seconds =======================================
```

### Show the coverage statistics

```
$ coverage report -m
Name                           Stmts   Miss  Cover   Missing
------------------------------------------------------------
conda_mirror/__init__.py           3      0   100%
conda_mirror/conda_mirror.py     236     20    92%   203-205, 209-210, 214, 240, 249-254, 262-264, 303, 366, 497, 542-543, 629
------------------------------------------------------------
TOTAL                            239     20    92%
```

## Other

After a new contributor makes a pull-request that is approved, we will reach out
and invite you to be a maintainer of the project.


## Releasing

To release you need three things

1. Commit rights to conda-mirror
2. A github token
3. The version number that you want to use for the new tag

After you have all three of these things, run the release.sh script (on a unix machine) and
pass it the tag that you want to use and your github token:

```bash
GITHUB_TOKEN=<github_token> ./release.sh <tag>
```


CHANGELOG.md
# Change Log

## [0.8.0](https://github.com/maxpoint/conda-mirror/tree/0.8.0)

- Improve validation performance [\#69](https://github.com/maxpoint/conda-mirror/issues/69)
- Check local disk space available [\#70](https://github.com/maxpoint/conda-mirror/issues/70)
- Log download errors instead of breaking and clearing [\#71](https://github.com/maxpoint/conda-mirror/issues/71)
- Add command line switch --no-validate-target [\#72](https://github.com/maxpoint/conda-mirror/issues/72)
- Support most CLI options in the config file [\#74](https://github.com/maxpoint/conda-mirror/issues/74)
- Correct numbering in output when validating [\#75](https://github.com/maxpoint/conda-mirror/issues/75)

[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.7.4...0.8.0)

## [0.7.4](https://github.com/maxpoint/conda-mirror/tree/0.7.4)

[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.7.3...0.7.4)

**Implemented enhancements:**

- Add a --dry-run mode [\#39](https://github.com/maxpoint/conda-mirror/issues/39)

**Merged pull requests:**

- ENH: First pass at dry-run [\#59](https://github.com/maxpoint/conda-mirror/pull/59) ([ericdill](https://github.com/ericdill))

## [0.7.3](https://github.com/maxpoint/conda-mirror/tree/0.7.3) (2017-08-07)
[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.7.2...0.7.3)

**Fixed bugs:**

- Keep old packages [\#56](https://github.com/maxpoint/conda-mirror/issues/56)

**Closed issues:**

- Validation errors caused by incomplete tar file [\#57](https://github.com/maxpoint/conda-mirror/issues/57)

**Merged pull requests:**

- Handle EOFError when reading tar files. [\#58](https://github.com/maxpoint/conda-mirror/pull/58) ([dmkent](https://github.com/dmkent))
- Add release script that edits the output of github\_changelog\_generator [\#55](https://github.com/maxpoint/conda-mirror/pull/55) ([ericdill](https://github.com/ericdill))
- DOC: Rerun changelog generator [\#54](https://github.com/maxpoint/conda-mirror/pull/54) ([ericdill](https://github.com/ericdill))

## [0.7.2](https://github.com/maxpoint/conda-mirror/tree/0.7.2) (2017-05-23)
[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.7.1...0.7.2)

**Merged pull requests:**

- Modify run summary. Configure flake8 on travis [\#53](https://github.com/maxpoint/conda-mirror/pull/53) ([ericdill](https://github.com/ericdill))
- MNT: Template out the message [\#52](https://github.com/maxpoint/conda-mirror/pull/52) ([ericdill](https://github.com/ericdill))
- Raise don't exit when required args aren't present [\#51](https://github.com/maxpoint/conda-mirror/pull/51) ([ericdill](https://github.com/ericdill))

## [0.7.1](https://github.com/maxpoint/conda-mirror/tree/0.7.1) (2017-05-18)
[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.7.0...0.7.1)

**Merged pull requests:**

- Capture information on downloaded and removed packages [\#50](https://github.com/maxpoint/conda-mirror/pull/50) ([ericdill](https://github.com/ericdill))

## [0.7.0](https://github.com/maxpoint/conda-mirror/tree/0.7.0) (2017-05-11)
[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.6.6...0.7.0)

**Closed issues:**

- concurrent package validation [\#43](https://github.com/maxpoint/conda-mirror/issues/43)

**Merged pull requests:**

- Add concurrent package validation [\#48](https://github.com/maxpoint/conda-mirror/pull/48) ([willirath](https://github.com/willirath))
- TST: Fix broken test [\#46](https://github.com/maxpoint/conda-mirror/pull/46) ([ericdill](https://github.com/ericdill))

## [0.6.6](https://github.com/maxpoint/conda-mirror/tree/0.6.6) (2017-03-28)
[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.6.5...0.6.6)

**Implemented enhancements:**

- Sort packages for validation [\#34](https://github.com/maxpoint/conda-mirror/issues/34)

**Closed issues:**

- Does not run under py 2.7 [\#42](https://github.com/maxpoint/conda-mirror/issues/42)
- Package license [\#36](https://github.com/maxpoint/conda-mirror/issues/36)
- setup.py home [\#35](https://github.com/maxpoint/conda-mirror/issues/35)

**Merged pull requests:**

- MNT/DOC: Make it clear this is py3 only [\#44](https://github.com/maxpoint/conda-mirror/pull/44) ([ericdill](https://github.com/ericdill))
- Add syntax highlight to the repodata example in README [\#41](https://github.com/maxpoint/conda-mirror/pull/41) ([nicoddemus](https://github.com/nicoddemus))
- Sort the validation output [\#40](https://github.com/maxpoint/conda-mirror/pull/40) ([ericdill](https://github.com/ericdill))
- Fix home to point to MaxPoint org [\#38](https://github.com/maxpoint/conda-mirror/pull/38) ([jakirkham](https://github.com/jakirkham))
- Package standard docs [\#37](https://github.com/maxpoint/conda-mirror/pull/37) ([jakirkham](https://github.com/jakirkham))

## [0.6.5](https://github.com/maxpoint/conda-mirror/tree/0.6.5) (2017-02-23)
[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.6.4...0.6.5)

**Closed issues:**

- FileNotFoundError after package is removed for failing size validation [\#31](https://github.com/maxpoint/conda-mirror/issues/31)

**Merged pull requests:**

- Return after removing the package [\#32](https://github.com/maxpoint/conda-mirror/pull/32) ([ericdill](https://github.com/ericdill))

## [0.6.4](https://github.com/maxpoint/conda-mirror/tree/0.6.4) (2017-02-23)
[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.6.3...0.6.4)

**Closed issues:**

- TypeError: 'set' object is not subscriptable [\#29](https://github.com/maxpoint/conda-mirror/issues/29)

**Merged pull requests:**

- Fix Type Error [\#30](https://github.com/maxpoint/conda-mirror/pull/30) ([ericdill](https://github.com/ericdill))

## [0.6.3](https://github.com/maxpoint/conda-mirror/tree/0.6.3) (2017-02-15)
[Full Changelog](https://github.com/maxpoint/conda-mirror/compare/0.6.2...0.6.3)

**Closed issues:**

- Mirror pro repo [\#24](https://github.com/maxpoint/conda-mirror/issues/24)


conda_mirror/cli.py
from __future__ import (unicode_literals, print_function, division,
                        absolute_import)
import argparse
import logging
import yaml
frin .conda_mirror import main

logging.basicConfig(level=logging.INFO,
                    format='%(asctime)-15s %(message)s')

DEFAULT_BAD_LICENSES = ['agpl', '']
REPODATA = 'https://conda.anaconda.org/{channel}/{platform}/repodata.json'
DEFAULT_PLATFORMS = ['linux-64',
                     'linux-32',
                     'osx-64',
                     'win-64',
                     'win-32']


def _make_arg_parser():
    """
    Localize the ArgumentParser logic

    Returns
    -------
    argument_parser : argparse.ArgumentParser
        The instantiated argument parser for this CLI
    """
    ap = argparse.ArgumentParser(description="CLI interface for conda-mirror.py")

    ap.add_argument(
        '--upstream-channel',
        help='The anaconda channel to mirror',
        required=True
    )
    ap.add_argument(
        '--target-directory',
        help='The place where packages should be mirrored to',
        required=True
    )
    ap.add_argument(
        '--platform',
        nargs="+",
        help=("The OS platform(s) to mirror. one or more of: {'all',"
              " 'linux-64', 'linux-32', 'osx-64', 'win-32', 'win-64'}"),
        required=True
    )
    ap.add_argument(
        '-v', '--verbose',
        action="store_true",
        help="This basically turns on tqdm progress bars for downloads",
        default=False,
    )
    ap.add_argument(
        '--config',
        action="store",
        help="Path to the yaml config file",
    )
    ap.add_argument(
        '--blacklist',
        action="+",
        help="k:v pairs to match against to then blacklist from the mirroirng",
        default=[]
    )
    ap.add_argument(
        '--whitelist',
        action="+",
        help=("k:v pairs to match against to then whitelist from the mirroring."
              "Note that whitelist supercedes blacklist"),
        default=[]
    )
    return ap


def cli():
    """
    Collect arguments from sys.argv and invoke the main() function.
    """
    ap = _make_arg_parser()
    args = ap.parse_args()

    if args.config:
        with open(args.config, 'r') as f:
            config = yaml

    main(args.upstream_channel, args.target_directory, args.platform, args.verbose)

